### DISCUSSION TITLE
Is there some preferred order for properties in quest class?

### DISCUSSION THREAD

**Discussion by matkoniecz** [streetcomplete/StreetComplete#3526] [Q&A] [answered]
Seems to be: 

- `elementFilter` (sometimes quest specific ElementFilterExpression)
- `changesetComment`
- `wikiLink`
- `icon`
- `isDeleteElementEnabled` (may be skipped)
- `hasMarkersAtEnds` (may be skipped)
- `isReplaceShopEnabled`  (may be skipped)
- `achievements`
- `defaultDisabledMessage` (may be skipped)
- `getTitle`
- `getApplicableElements`  (may be skipped)
- `isApplicableTo`  (may be skipped)
- `getHighlightedElements`  (may be skipped)
- `createForm`
- `applyAnswerTo`

based on what seems popular.

See `rg "val (elementFilter|changesetComment|wikiLink|icon|isDeleteElementEnabled|hasMarkersAtEnds|isReplaceShopEnabled|achievements|getTitle|getApplicableElements|isApplicableTo|getHighlightedElements|createForm|applyAnswerTo|defaultDisabledMessage)" --only-matching`

This order is not universal, but most more or less follow it what is confusing a bit.

I do not care at all which order is used but I am tempted to reorder some classes for consistency so that in future I can check any class and see the preferred order.

Or is it not necessary to care about consistency here and each quest may have a different one? Right now I am reordering one in my WIP because I guess that answer here is "no".

EDIT: commitMessage changed to changesetComment, questTypeAchievements to achievements, isSplitWayEnabled removed to keep up with code changes. And added defaultDisabledMessage per https://github.com/streetcomplete/StreetComplete/discussions/3526#discussioncomment-3349206 https://github.com/streetcomplete/StreetComplete/discussions/3526#discussioncomment-3348984

**Comment by FloEdelmann** [comment]
I always like consistency, so I'd prefer one PR to update all quest type classes to have the same order. The order you suggested seems fine to me, but I don't care about the specific order too much.

**Comment by smichel17** [comment]
The [kotin coding conventions](https://kotlinlang.org/docs/coding-conventions.html#class-layout) recommend 

> The contents of a class should go in the following order:
> 
> 1. Property declarations and initializer blocks
> 2. Secondary constructors
> 3. Method declarations
> 4. Companion object

Past that, while I don't have a strong opinion, I'd encourage you to consider **alphabetical order**. It adds a small amount of friction the first time you encounter a class (vs logical order), but removes the mental overhead of remembering the order and is totally unambiguous.

**Reply by matkoniecz** [reply]
I think that `icon` `wikiLink` `commitMessage` really should be together. Though I would also keep it together with also `getTitle`

Similar to `getApplicableElements` and `isApplicableTo`. But I would put them with `elementFilter`.

Hmmm. If I would be rearranging it anyway I may actually do this. And put `getApplicableElements` and `isApplicableTo` next to `elementFilter`.

**Comment by westnordost** [comment, marked as answer]
I have a very rough preferred order, not sure if it is done consistently in all quests:

First basic properties (filter, etc. including title and icon), then roughly in the order it is used - any quest filter logic, then creating the form, then applying the change from the answer. Companion object and global functions/fields are at the end

**Reply by matkoniecz** [reply]
> including title

You mean `commitMessage` not `getTitle`, right?

**Reply by matkoniecz** [reply]
@westnordost What about `questTypeAchievements`? Is it considered as basic property?

According to you answer it seems to be more fitting class of less important properties and put at the end as acting after user answered. But in most quests it is after `icon`.  

Similarly `isDeleteElementEnabled` - is it also supposed to be treated as a basic property listed just after icon?

Asking as I plan to implement a bit of unification based on this discussion.

**Reply by westnordost** [reply]
Yeah, sounds good

**Reply by matkoniecz** [reply]
@westnordost Now I am confused which version is good :)

I plan to use https://github.com/streetcomplete/StreetComplete/discussions/3526#discussion-3699222 order - `isDeleteElementEnabled` and `getTitle` as a regular part of form definition, `questTypeAchievements` listed near end.

Is it OK?

**Reply by westnordost** [reply]
The order listed below "Seems to be:" in the linked comment is good

**Comment by matkoniecz** [comment]
@westnordost Do you have any preference where `defaultDisabledMessage` should go? https://github.com/streetcomplete/StreetComplete/blob/7c97885cb6c24f9f7defed22e4497a495ab7e4ab/app/src/main/java/de/westnordost/streetcomplete/quests/width/AddCyclewayWidth.kt#L39 that you authored has it after achievements, for example which seems entirely fine to me.

Though it often is listed before achievements. And in few cases somewhere else.

I have absolutely no preferences here, plan on adding it on recommended list to be listed after achievements and reformat code to achieve that to have one less thing to decide on when writing code.

**Comment by westnordost** [comment]
No

On August 8, 2022 2:13:27 PM GMT+02:00, Mateusz Konieczny ***@***.***> wrote:
***@***.*** Do you have any preference where `defaultDisabledMessage` should go? https://github.com/streetcomplete/StreetComplete/blob/7c97885cb6c24f9f7defed22e4497a495ab7e4ab/app/src/main/java/de/westnordost/streetcomplete/quests/width/AddCyclewayWidth.kt#L39 that you authored has it after achievements, for example which seems entirely fine to me.
>
>Though it often is listed before achievements. And in few cases somewhere else.
>
>I have absolutely no preferences here, plan on adding it on recommended list to be listed after achievements and reformat code to achieve that to have one less thing to decide on when writing code.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/streetcomplete/StreetComplete/discussions/3526#discussioncomment-3348984
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>
-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.
